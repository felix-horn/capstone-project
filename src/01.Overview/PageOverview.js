import { useState } from 'react'
import PropTypes from 'prop-types'
import styled from 'styled-components/macro'

import ShopCard from './02.Components/ShopCard'
import OverlayNavigation from './02.Components/OverlayNavigation'
import ButtonCircle from './01.UI-Elements/ButtonCircle'
import MenuIcon from '@material-ui/icons/Menu'

OverviewPage.propTypes = {
  database: PropTypes.object.isRequired,
  addShop: PropTypes.func.isRequired,
}

export default function OverviewPage({ database, addShop }) {
  const [isOverlayVisible, setIsOverlayVisible] = useState(false)
  const [buttonNavigationClass, setButtonNavigationClass] = useState('')
  const allShopIds = database.shops.allIds
  return (
    <PageLayout>
      {isOverlayVisible && (
        <OverlayNavigationPositioned
          onClick={handleOverlayClick}
          addShop={addShop}
        />
      )}
      {allShopIds.map((shopId) => (
        <ShopCard key={shopId} shopId={shopId} database={database} />
      ))}
      <ButtonNavigation
        onClick={handleButtonNavigationClick}
        className={buttonNavigationClass}
      >
        <MenuIconAnimated className={buttonNavigationClass} />
      </ButtonNavigation>
    </PageLayout>
  )

  function handleButtonNavigationClick() {
    setIsOverlayVisible(true)
    setButtonNavigationClass('')
  }

  function handleOverlayClick() {
    setIsOverlayVisible(false)
    setButtonNavigationClass('rotate-back')
  }
}

const PageLayout = styled.div`
  height: calc(100vh - 50px - 50px); // GlobalStyles: 50px top and bottom
  display: grid;
  gap: 10px;
  grid-auto-rows: minmax(min-content, max-content);
`
const OverlayNavigationPositioned = styled(OverlayNavigation)`
  position: fixed;
  z-index: var(--z-index-overlay);
  top: 0;
  left: 0;
`
const ButtonNavigation = styled(ButtonCircle)`
  position: fixed;
  z-index: var(--z-index-floating-item);
  bottom: 170px;
  right: 40px;
`
const MenuIconAnimated = styled(MenuIcon)`
  &.rotate-back {
    animation: 0.2s rotateBack ease-in-out;

    @keyframes rotateBack {
      0% {
        transform: rotate(90deg);
      }
      100% {
        transform: rotate(0deg);
      }
    }
  }
`
