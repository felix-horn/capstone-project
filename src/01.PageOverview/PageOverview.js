import { useState } from 'react'
import PropTypes from 'prop-types'
import styled from 'styled-components/macro'
import ShopCard from './02.Components/ShopCard'
import ButtonNavigation from './01.UI-Elements/ButtonNavigation'
import OverlayNavigation from './02.Components/OverlayNavigation'

OverviewPage.propTypes = {
  database: PropTypes.object.isRequired,
  addShop: PropTypes.func.isRequired,
}

export default function OverviewPage({ database, addShop }) {
  const [isOverlayVisible, setIsOverlayVisible] = useState(false)
  const [buttonNavigationClass, setButtonNavigationClass] = useState('')
  return (
    <PageGrid>
      {isOverlayVisible && (
        <OverlayNavigationPositioned
          onClick={handleOverlayClick}
          addShop={addShop}
        />
      )}
      {database.shops.allIds.map((shopId) => (
        <ShopCard key={shopId} shopId={shopId} database={database} />
      ))}
      <ButtonNavigationPositioned
        onClick={handleButtonNavigationClick}
        className={buttonNavigationClass}
      />
    </PageGrid>
  )

  function handleButtonNavigationClick() {
    setIsOverlayVisible(true)
    setButtonNavigationClass('')
  }

  function handleOverlayClick() {
    setIsOverlayVisible(false)
    setButtonNavigationClass('rotate-back')
  }
}

const PageGrid = styled.div`
  position: relative;
  height: calc(100vh - 50px - 10px); // 50px top and 10px bottom
  display: grid;
  gap: 10px;
  grid-auto-rows: minmax(min-content, max-content);
`
const OverlayNavigationPositioned = styled(OverlayNavigation)`
  position: fixed;
  z-index: var(--z-index-overlay);
  top: 0;
  left: 0;
`
const ButtonNavigationPositioned = styled(ButtonNavigation)`
  position: absolute;
  bottom: 60px;
  right: 30px;
  z-index: var(--z-index-floating-item);
`
