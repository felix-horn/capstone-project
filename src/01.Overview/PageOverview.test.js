import { render } from '@testing-library/react'
import user from '@testing-library/user-event'
import { MemoryRouter as Router } from 'react-router-dom'
import PageOverview from './PageOverview'

const testDatabase = {
  shops: {
    allIds: ['x', 'y'],
    byId: {
      x: {
        id: 'x',
        title: 'shopX',
        items: ['a', 'b'],
      },
      y: {
        id: 'y',
        title: 'shopY',
        items: ['c'],
      },
    },
  },
  items: {
    allIds: ['a', 'b', 'c'],
    byId: {
      a: {
        id: 'a',
        title: 'itemA',
        isChecked: false,
      },
      b: {
        id: 'b',
        title: 'itemB',
        isChecked: true,
      },
      c: {
        id: 'c',
        title: 'itemC',
        isChecked: false,
      },
    },
  },
}

/* export as global helper function */

function renderWithRouter(children) {
  return render(<Router>{children}</Router>)
}

describe('PageOverview', () => {
  it('renders correctly', () => {
    const addShopMock = jest.fn()
    const { container } = renderWithRouter(
      <PageOverview database={testDatabase} addShop={addShopMock} />
    )
    expect(container.firstChild).toMatchSnapshot()
  })

  it('calls on addShop', () => {
    const addShopMock = jest.fn()
    const { getByTestId } = render(
      <Router>
        <PageOverview database={testDatabase} addShop={addShopMock} />
      </Router>
    )
    const button = getByTestId('action-button')
    user.click(button)
    expect(addShopMock).toHaveBeenCalledTimes(1)
  })

  it('links to correct href', () => {
    const addShopMock = jest.fn()
    const { getByTestId } = render(
      <Router>
        <PageOverview database={testDatabase} addShop={addShopMock} />
      </Router>
    )
    const button = getByTestId('action-button')
    user.click(button)
    const href = button.href
    expect(href).toMatch('/shop')
  })
})
