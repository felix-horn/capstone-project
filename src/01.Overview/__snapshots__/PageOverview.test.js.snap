// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PageOverview renders correctly 1`] = `
.c2 {
  margin-bottom: 8px !important;
  font-size: 1.2rem;
  font-weight: 400;
  color: var(--almost-black);
}

.c3 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  gap: 5px;
}

.c4 {
  border: 1px solid var(--light-gray);
  border-radius: 2px;
  height: 15px;
  width: 15px;
}

.c5 {
  margin: 0;
  color: var(--dark-gray);
}

.c6 {
  margin: 4px 2px 0 !important;
  padding: 0;
  font-size: 0.8rem;
  color: var(--light-gray);
}

.c1 {
  border: var(--border);
  border-radius: 8px;
  padding: 15px;
  display: grid;
  gap: 5px;
  -webkit-text-decoration: none;
  text-decoration: none;
}

.c7 {
  box-shadow: var(--strong-box-shadow);
  border-radius: 100%;
  height: 50px;
  width: 50px;
  background-color: var(--white);
  display: grid;
  place-items: center;
}

.c9 {
  position: relative !important;
  bottom: 0px !important;
  right: 0px !important;
  color: var(--CTA-blue);
}

.c9.rotate-back {
  -webkit-animation: 0.2s rotateBack ease-in-out;
  animation: 0.2s rotateBack ease-in-out;
}

.c0 {
  position: relative;
  height: calc(100vh - 50px - 10px);
  display: grid;
  gap: 10px;
  grid-auto-rows: minmax(min-content,max-content);
}

.c8 {
  position: absolute;
  bottom: 60px;
  right: 30px;
  z-index: var(--z-index-floating-item);
}

<div
  class="c0"
>
  <a
    class="c1"
    data-testid="shop-card"
    href="/shop/shopX"
  >
    <h2
      class="c2"
    >
      shopX
    </h2>
    <label
      class="c3"
    >
      <div
        class="c4"
      />
      <p
        class="c5"
      >
        itemA
      </p>
    </label>
    <p
      class="c6"
    >
      und 1 abgehakter Eintrag
    </p>
  </a>
  <a
    class="c1"
    data-testid="shop-card"
    href="/shop/shopY"
  >
    <h2
      class="c2"
    >
      shopY
    </h2>
    <label
      class="c3"
    >
      <div
        class="c4"
      />
      <p
        class="c5"
      >
        itemC
      </p>
    </label>
  </a>
  <div
    class="c7 c8"
  >
    <svg
      aria-hidden="true"
      class="MuiSvgIcon-root c9 c8"
      focusable="false"
      viewBox="0 0 24 24"
    >
      <path
        d="M3 18h18v-2H3v2zm0-5h18v-2H3v2zm0-7v2h18V6H3z"
      />
    </svg>
  </div>
</div>
`;
